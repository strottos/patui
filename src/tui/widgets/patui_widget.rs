use std::fmt::Debug;

use ratatui::{buffer::Buffer, layout::Rect, widgets::WidgetRef};

use super::{text::Text, ScrollableArea, Table};

#[cfg(test)]
pub(crate) use tests::TestWidget;

#[derive(Debug)]
pub(crate) enum ScrollType {
    Single(isize),
    HalfPageUp,
    HalfPageDown,
    HalfPageLeft,
    HalfPageRight,
    FullPageUp,
    FullPageDown,
}

#[derive(Clone, Debug)]
pub(crate) struct PatuiWidget<'a> {
    inner: PatuiWidgetInner<'a>,
}

impl<'a> PatuiWidget<'a> {
    pub(crate) fn new_text(text: Text<'a>) -> PatuiWidget<'a> {
        Self {
            inner: PatuiWidgetInner::Text(text),
        }
    }

    pub(crate) fn new_scrollable_area(scrollable_area: ScrollableArea<'a>) -> PatuiWidget<'a> {
        Self {
            inner: PatuiWidgetInner::ScrollableArea(scrollable_area),
        }
    }

    pub(crate) fn new_table(table: Table<'a>) -> PatuiWidget<'a> {
        Self {
            inner: PatuiWidgetInner::Table(table),
        }
    }

    pub(crate) fn scrollable_height(&self) -> usize {
        match &self.inner {
            PatuiWidgetInner::Text(text) => text.height(),
            // PatuiWidget::Button(button) => button.scrollable_height(),
            PatuiWidgetInner::ScrollableArea(_) => todo!(),
            PatuiWidgetInner::Table(table) => table.scrollable_height(),
            #[cfg(test)]
            PatuiWidgetInner::TestWidget(test_widget) => test_widget.height,
        }
    }

    pub(crate) fn num_widgets(&self) -> usize {
        match &self.inner {
            PatuiWidgetInner::Text(_) => 1,
            // PatuiWidget::Button(button) => button.num_widgets(),
            PatuiWidgetInner::ScrollableArea(_) => todo!(),
            PatuiWidgetInner::Table(table) => table.num_elements(),
            #[cfg(test)]
            PatuiWidgetInner::TestWidget(test_widget) => test_widget.num_widgets,
        }
    }

    pub(crate) fn is_selectable(&self) -> bool {
        match &self.inner {
            PatuiWidgetInner::Text(text) => text.is_selectable(),
            PatuiWidgetInner::ScrollableArea(_) => todo!(),
            PatuiWidgetInner::Table(_) => true,
            #[cfg(test)]
            PatuiWidgetInner::TestWidget(_) => false,
        }
    }

    pub(crate) fn set_selected(&self, selected: bool) -> &Self {
        match &self.inner {
            PatuiWidgetInner::Text(text) => {
                text.set_selected(selected);
            }
            PatuiWidgetInner::ScrollableArea(_) => todo!(),
            PatuiWidgetInner::Table(_) => {
                // table.set_selected(selected);
            }
            #[cfg(test)]
            PatuiWidgetInner::TestWidget(_) => {}
        }

        self
    }

    pub(crate) fn inner_scrollable(&self) -> &ScrollableArea<'a> {
        match &self.inner {
            PatuiWidgetInner::ScrollableArea(scrollable_area) => scrollable_area,
            _ => panic!("Cannot get reference to non-scrollable area"),
        }
    }

    pub(crate) fn inner_scrollable_mut(&mut self) -> Option<&mut ScrollableArea<'a>> {
        match &mut self.inner {
            PatuiWidgetInner::ScrollableArea(scrollable_area) => Some(scrollable_area),
            _ => None,
        }
    }

    pub(crate) fn inner_table_mut(&mut self) -> Option<&mut Table<'a>> {
        match &mut self.inner {
            PatuiWidgetInner::Table(table) => Some(table),
            _ => None,
        }
    }

    #[cfg(test)]
    pub(crate) fn new_test(test: TestWidget<'a>) -> PatuiWidget<'a> {
        Self {
            inner: PatuiWidgetInner::TestWidget(test),
        }
    }

    #[cfg(test)]
    pub(crate) fn get_test_inner(&self) -> Option<&TestWidget> {
        match &self.inner {
            PatuiWidgetInner::TestWidget(test_widget) => Some(test_widget),
            _ => None,
        }
    }
}

impl<'a> WidgetRef for PatuiWidget<'a> {
    fn render_ref(&self, area: Rect, buf: &mut Buffer) {
        match &self.inner {
            PatuiWidgetInner::Text(text) => text.render_ref(area, buf),
            // PatuiWidgetInner::Button(button) => button.render_ref(area, buf),
            PatuiWidgetInner::ScrollableArea(scrollable_area) => {
                scrollable_area.render_ref(area, buf)
            }
            PatuiWidgetInner::Table(table) => table.render_ref(area, buf),
            #[cfg(test)]
            PatuiWidgetInner::TestWidget(test_widget) => test_widget.render_ref(area, buf),
        }
    }
}

#[derive(Clone, Debug)]
pub(crate) enum PatuiWidgetInner<'a> {
    Text(Text<'a>),
    // Button(Button),
    ScrollableArea(ScrollableArea<'a>),
    Table(Table<'a>),
    #[cfg(test)]
    TestWidget(TestWidget<'a>),
}

#[cfg(test)]
mod tests {
    use std::cell::RefCell;

    use ratatui::{buffer::Buffer, layout::Rect, text::Text, widgets::WidgetRef};

    #[derive(Clone, Debug)]
    pub(crate) struct TestWidget<'a> {
        pub height: usize,
        pub num_widgets: usize,
        pub calls: RefCell<Vec<Rect>>,
        pub text: Text<'a>,
    }

    impl<'a> TestWidget<'a> {
        pub(crate) fn new(height: usize, num_widgets: usize) -> Self {
            Self {
                height,
                num_widgets,
                calls: RefCell::new(vec![]),
                text: Text::from(""),
            }
        }

        pub(crate) fn new_with_text(
            height: usize,
            num_widgets: usize,
            text: Text<'a>,
        ) -> TestWidget<'a> {
            Self {
                height,
                num_widgets,
                calls: RefCell::new(vec![]),
                text,
            }
        }
    }

    impl<'a> WidgetRef for TestWidget<'a> {
        fn render_ref(&self, area: Rect, buf: &mut Buffer) {
            self.calls.borrow_mut().push(area);
            self.text.render_ref(area, buf);
        }
    }
}
