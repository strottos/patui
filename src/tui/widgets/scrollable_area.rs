use std::{cell::Cell, cmp};

use ratatui::{
    buffer::Buffer,
    layout::Rect,
    prelude::BlockExt,
    widgets::{Block, Scrollbar, ScrollbarOrientation, ScrollbarState, StatefulWidget, WidgetRef},
};
use tracing::{debug, trace};

use super::patui_widget::PatuiWidget;

#[derive(Debug)]
pub(crate) struct ScrollableArea<'a> {
    first_row: isize,
    selected_rows: (isize, isize),
    display_height: Cell<usize>,
    widgets: Vec<PatuiWidget<'a>>,
    block: Option<Block<'a>>,
}

impl<'a> ScrollableArea<'a> {
    pub(crate) fn new_patui_widget() -> PatuiWidget<'a> {
        PatuiWidget::ScrollableArea(Self {
            first_row: 0,
            selected_rows: (-1, -1),
            display_height: Cell::new(24),
            widgets: vec![],
            block: None,
        })
    }

    pub(crate) fn add_widget(&mut self, widget: PatuiWidget<'a>) -> &mut Self {
        self.widgets.push(widget);
        self
    }

    pub(crate) fn set_widgets(&mut self, widgets: Vec<PatuiWidget<'a>>) -> &mut Self {
        self.widgets = widgets.into_iter().map(|widget| widget).collect();
        self
    }

    pub(crate) fn add_block(&mut self, block: Block<'a>) -> &mut Self {
        self.block = Some(block);
        self
    }

    /// Alters the selected_line by the count given with optional wrapping
    /// and if necessary scrolls the page to ensure the selected_line is
    /// visible.
    ///
    /// Returns true if changed else false
    pub(crate) fn navigate(&mut self, scroll_lines: isize, wrap_around: bool) {
        todo!()
    }

    pub(crate) fn scroll(&mut self, scroll_lines: isize) {
        self.first_row += scroll_lines;
        self.first_row = cmp::max(0, self.first_row);
        self.first_row = cmp::min(
            self.get_total_height() as isize - self.display_height.get() as isize,
            self.first_row,
        );
    }

    /// Renders the widgets in the scrollale area that should be visible.
    fn render_main(&self, area: Rect, buf: &mut Buffer) {
        let area_height = area.height as usize;

        self.display_height.set(area_height);

        let mut current_height = 0;
        let mut current_row = 0;

        for widget in &self.widgets {
            let widget_height = widget.scrollable_height();

            let skip_lines = if current_row < self.first_row as usize {
                self.first_row as usize - current_row
            } else {
                0
            };
            current_row += widget_height;

            if current_height + widget_height <= self.first_row as usize {
                continue;
            }

            let widget_area = Rect {
                x: area.x,
                y: area.y + current_height as u16,
                width: area.width,
                height: cmp::min(area_height - current_height, widget_height - skip_lines) as u16,
            };

            widget
                .set_render_from_line(skip_lines)
                .render_ref(widget_area, buf);

            current_height += widget_height - skip_lines;

            if current_height >= area_height {
                break;
            }
        }
    }

    /// Renders the scrollbar on the right side of the area.
    fn render_scrollbar(&self, area: Rect, buf: &mut Buffer) {
        let scrollbar = Scrollbar::new(ScrollbarOrientation::VerticalRight)
            .begin_symbol(Some("↑"))
            .end_symbol(Some("↓"));

        let scrollbar_height = if self.get_total_height() <= self.display_height.get() {
            0
        } else {
            self.get_total_height() - self.display_height.get()
        };

        let mut scrollbar_state =
            ScrollbarState::new(scrollbar_height).position(self.first_row as usize);

        scrollbar.render(area, buf, &mut scrollbar_state);
    }

    fn get_total_height(&self) -> usize {
        self.widgets
            .iter()
            .map(|widget| widget.scrollable_height())
            .sum()
    }
}

impl<'a> WidgetRef for ScrollableArea<'a> {
    fn render_ref(&self, area: Rect, buf: &mut Buffer) {
        debug!("Rendering scrollable area: {:?}", area);
        self.block.render_ref(area, buf);
        let inner = self.block.inner_if_some(area);
        let inner_main = Rect {
            x: inner.x,
            y: inner.y,
            width: inner.width - 1,
            height: inner.height,
        };
        self.render_main(inner_main, buf);
        let scrollbar_rect = if self.block.is_some() {
            // NB: The scrollbar is rendered on the right side of the block, we have to make
            // assumptions about the border so far, or more work needed.
            Rect {
                x: area.x + area.width - 2,
                y: area.y + 1,
                width: 1,
                height: area.height - 2,
            }
        } else {
            Rect {
                x: area.width - 1,
                y: area.y,
                width: 1,
                height: area.height,
            }
        };
        trace!("Rendering scrollbar rect: {:?}", scrollbar_rect);
        self.render_scrollbar(scrollbar_rect, buf);
        trace!("Rendered scrollable area: {:?}", buf);
    }
}

// impl<'a> PatuiWidget for ScrollableArea<'a> {
// fn scrollable_height(&self) -> usize {
//     self.widgets
//         .iter()
//         .map(|widget| widget.scrollable_height())
//         .sum()
// }

// fn num_widgets(&self) -> usize {
//     self.widgets.iter().map(|widget| widget.num_widgets()).sum()
// }
// }

#[cfg(test)]
mod tests {
    use crate::tui::widgets::patui_widget::TestWidget;

    use super::*;
    use assertor::*;
    use ratatui::widgets::{Borders, Padding};

    fn get_widget_calls(scrollable_area: &ScrollableArea) -> Vec<TestWidget> {
        scrollable_area
            .widgets
            .iter()
            .map(|widget| widget.get_test_inner().unwrap().clone())
            .collect::<Vec<_>>()
    }

    #[test]
    fn test_scrollable_area_simple() {
        let widgets = vec![
            PatuiWidget::TestWidget(TestWidget::new(3, 1)),
            PatuiWidget::TestWidget(TestWidget::new(6, 1)),
        ];
        let scrollable_area = ScrollableArea {
            first_row: 0,
            selected_rows: (-1, -1),
            display_height: Cell::new(24),
            widgets,
            block: None,
        };
        let rect = Rect::new(0, 0, 10, 10);
        let mut buffer = Buffer::empty(rect);

        scrollable_area.render_ref(rect, &mut buffer);

        assert_that!(scrollable_area.display_height.get()).is_equal_to(10);

        let test_widgets = get_widget_calls(&scrollable_area);
        assert_that!(test_widgets).has_length(2);
        assert_that!(*test_widgets[0].calls.borrow()).has_length(1);
        assert_that!(test_widgets[0].calls.borrow()[0]).is_equal_to(Rect::new(0, 0, 9, 3));
        assert_that!(test_widgets[0].render_from_line.get()).is_equal_to(0);
        assert_that!(*test_widgets[1].calls.borrow()).has_length(1);
        assert_that!(test_widgets[1].calls.borrow()[0]).is_equal_to(Rect::new(0, 3, 9, 6));
        assert_that!(test_widgets[1].render_from_line.get()).is_equal_to(0);

        insta::assert_debug_snapshot!(buffer);
    }

    #[test]
    fn test_scrollable_area_with_overlaps() {
        let widgets = vec![
            PatuiWidget::TestWidget(TestWidget::new(3, 1)),
            PatuiWidget::TestWidget(TestWidget::new(6, 1)),
            PatuiWidget::TestWidget(TestWidget::new(7, 1)),
            PatuiWidget::TestWidget(TestWidget::new(3, 1)),
        ];
        let scrollable_area = ScrollableArea {
            first_row: 0,
            selected_rows: (-1, -1),
            display_height: Cell::new(24),
            widgets,
            block: None,
        };
        let rect = Rect::new(0, 0, 10, 10);
        let mut buffer = Buffer::empty(rect);

        scrollable_area.render_ref(rect, &mut buffer);

        let test_widgets = get_widget_calls(&scrollable_area);
        assert_that!(test_widgets).has_length(4);
        assert_that!(*test_widgets[0].calls.borrow()).has_length(1);
        assert_that!(test_widgets[0].calls.borrow()[0]).is_equal_to(Rect::new(0, 0, 9, 3));
        assert_that!(test_widgets[0].render_from_line.get()).is_equal_to(0);
        assert_that!(*test_widgets[1].calls.borrow()).has_length(1);
        assert_that!(test_widgets[1].calls.borrow()[0]).is_equal_to(Rect::new(0, 3, 9, 6));
        assert_that!(test_widgets[1].render_from_line.get()).is_equal_to(0);
        assert_that!(*test_widgets[2].calls.borrow()).has_length(1);
        assert_that!(test_widgets[2].calls.borrow()[0]).is_equal_to(Rect::new(0, 9, 9, 1));
        assert_that!(test_widgets[2].render_from_line.get()).is_equal_to(0);
        assert_that!(*test_widgets[3].calls.borrow()).has_length(0);

        insta::assert_debug_snapshot!(buffer);
    }

    #[test]
    fn test_scrollable_area_with_different_first_line() {
        let widgets = vec![
            PatuiWidget::TestWidget(TestWidget::new(3, 1)),
            PatuiWidget::TestWidget(TestWidget::new(6, 1)),
            PatuiWidget::TestWidget(TestWidget::new(7, 1)),
            PatuiWidget::TestWidget(TestWidget::new(3, 1)),
        ];
        let scrollable_area = ScrollableArea {
            first_row: 5,
            selected_rows: (-1, -1),
            display_height: Cell::new(24),
            widgets,
            block: None,
        };
        let rect = Rect::new(0, 0, 10, 10);
        let mut buffer = Buffer::empty(rect);

        scrollable_area.render_ref(rect, &mut buffer);

        let test_widgets = get_widget_calls(&scrollable_area);
        assert_that!(test_widgets).has_length(4);
        assert_that!(*test_widgets[0].calls.borrow()).has_length(0);
        assert_that!(*test_widgets[1].calls.borrow()).has_length(1);
        assert_that!(test_widgets[1].calls.borrow()[0]).is_equal_to(Rect::new(0, 0, 9, 4));
        assert_that!(test_widgets[1].render_from_line.get()).is_equal_to(2);
        assert_that!(*test_widgets[2].calls.borrow()).has_length(1);
        assert_that!(test_widgets[2].calls.borrow()[0]).is_equal_to(Rect::new(0, 4, 9, 6));
        assert_that!(test_widgets[2].render_from_line.get()).is_equal_to(0);
        assert_that!(*test_widgets[3].calls.borrow()).has_length(0);

        insta::assert_debug_snapshot!(buffer);
    }

    #[test]
    fn test_simple_scroll() {
        let widgets = (0..12)
            .map(|_| PatuiWidget::TestWidget(TestWidget::new(1, 1)))
            .collect::<Vec<_>>();
        let mut scrollable_area = ScrollableArea {
            first_row: 0,
            selected_rows: (-1, -1),
            display_height: Cell::new(5),
            widgets,
            block: None,
        };
        let rect = Rect::new(0, 0, 5, 5);
        let mut buffer = Buffer::empty(rect);

        scrollable_area.render_ref(rect, &mut buffer);
        scrollable_area.scroll(5);

        assert_that!(scrollable_area.first_row).is_equal_to(5);
        assert_that!(scrollable_area.selected_rows).is_equal_to((-1, -1));

        scrollable_area.scroll(5);

        assert_that!(scrollable_area.first_row).is_equal_to(7);
        assert_that!(scrollable_area.selected_rows).is_equal_to((-1, -1));

        scrollable_area.render_ref(rect, &mut buffer);
        insta::assert_debug_snapshot!(buffer);

        scrollable_area.scroll(5);

        assert_that!(scrollable_area.first_row).is_equal_to(7);
        assert_that!(scrollable_area.selected_rows).is_equal_to((-1, -1));

        scrollable_area.scroll(-5);

        assert_that!(scrollable_area.first_row).is_equal_to(2);
        assert_that!(scrollable_area.selected_rows).is_equal_to((-1, -1));

        scrollable_area.scroll(-5);

        assert_that!(scrollable_area.first_row).is_equal_to(0);
        assert_that!(scrollable_area.selected_rows).is_equal_to((-1, -1));

        scrollable_area.scroll(-5);

        assert_that!(scrollable_area.first_row).is_equal_to(0);
        assert_that!(scrollable_area.selected_rows).is_equal_to((-1, -1));
    }

    #[test]
    fn test_simple_render_block() {
        let widgets = vec![
            PatuiWidget::TestWidget(TestWidget::new(4, 1)),
            PatuiWidget::TestWidget(TestWidget::new(4, 1)),
            PatuiWidget::TestWidget(TestWidget::new(4, 1)),
        ];
        let scrollable_area = ScrollableArea {
            first_row: 0,
            selected_rows: (-1, -1),
            display_height: Cell::new(5),
            widgets,
            block: Some(
                Block::default()
                    .borders(Borders::ALL)
                    .padding(Padding::symmetric(2, 1)),
            ),
        };
        let rect = Rect::new(2, 2, 10, 10);
        let mut buffer = Buffer::empty(rect);

        scrollable_area.render_ref(rect, &mut buffer);

        let test_widgets = get_widget_calls(&scrollable_area);
        assert_that!(test_widgets).has_length(3);
        assert_that!(*test_widgets[0].calls.borrow()).has_length(1);
        assert_that!(test_widgets[0].calls.borrow()[0]).is_equal_to(Rect::new(5, 4, 3, 4));
        assert_that!(test_widgets[0].render_from_line.get()).is_equal_to(0);
        assert_that!(*test_widgets[1].calls.borrow()).has_length(1);
        assert_that!(test_widgets[1].calls.borrow()[0]).is_equal_to(Rect::new(5, 8, 3, 2));
        assert_that!(test_widgets[1].render_from_line.get()).is_equal_to(0);
        assert_that!(*test_widgets[2].calls.borrow()).has_length(0);

        insta::assert_debug_snapshot!(buffer);
    }
}
