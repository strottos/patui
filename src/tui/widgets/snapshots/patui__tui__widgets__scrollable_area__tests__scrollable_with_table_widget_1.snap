---
source: src/tui/widgets/scrollable_area.rs
expression: buffer
---
Buffer {
    area: Rect { x: 0, y: 0, width: 80, height: 24 },
    content: [
        "Test1                                                                          ↑",
        "Test2                                                                          █",
        "Test3                                                                          █",
        "Test4                                                                          █",
        "Test5                                                                          █",
        "Test6                                                                          █",
        "Test7                                                                          █",
        "Test8                                                                          █",
        "Test9                                                                          █",
        "Test10                                                                         █",
        "Test11                                                                         █",
        "Test12                                                                         █",
        "Test13                                                                         █",
        "Test14                                                                         █",
        "Test15                                                                         █",
        "Test16                                                                         █",
        "Test17                                                                         █",
        "Test18                                                                         █",
        "Test19                                                                         █",
        "Test20                                                                         ║",
        "Header1                    Header2                   Header3                   ║",
        "                                                                               ║",
        "Test1_1                    Test1_2                   Test1_3                   ║",
        "Test2_1                    Test2_2                   Test2_3                   ↓",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
